### ISSUE TITLE
support `unique_indices` option for `unique`

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#16330]
## 🚀 Feature
Requested by @gkioxari, it would bring `unique` api closer to numpy equivalent. Cuda implementation is not hard, but I haven't looked at CPU. 
May be #15804 scope should be widened to also include this. 
cc @ptrblck, @gkioxari, @jcjohnson

**Comment by jcjohnson** [comment]
This would be great! It is also related to https://github.com/pytorch/pytorch/issues/12598 -- if we have inverse indexes, unique indices, and counts (for both dim and non-dim) then we will be at feature parity with np.unique.

However for our use case, unique indices for single-dim on GPU will be perfect.

**Comment by ngimel** [comment]
Thanks for clarification.

**Comment by shaeqahmed** [comment]
Hi I'm a first time contributor and would like to work on this. Can you please give me some pointers / more context that would help me get started? Thanks!

**Comment by ngimel** [comment]
You need to add another option to unique API, you can use https://github.com/pytorch/pytorch/pull/15495/files as a blueprint for that. 
For GPU implementation, you can use thrust::unique_by_key to return unique_indices (this commit did that https://github.com/pytorch/pytorch/pull/16145/commits/883edac17355bc455f64fce7d4b8c93984eeae0a), I'm not sure about parallel CPU implementation but sequential CPU implementation should not be hard. 
When dim is not none it is more annoying, but generally similar thrust calls can be used on the GPU side, with corresponding comparison functors to compare vectors instead of single elements.

**Comment by zasdfgbnm** [comment]
I am assigning this to myself. I will work on this after working on https://github.com/pytorch/pytorch/issues/18987.

**Comment by OhadRubin** [comment]
What is the status of this issue??
Because right now the only solution is to use cupy and dlpack, since they offer this functionality.

**Comment by arbabu123** [comment]
it will be really useful to have this feature. please consider adding it :)

**Comment by ehsansaleh** [comment]
https://github.com/pytorch/pytorch/issues/36748 is also discussing a similar need. As a band-aid, one could consider using the following function. It doesn't require low-level manipulations neither in CUDA nor CPU. 

* For PyTorch Versions 1.13 and above (i.e., capable of stable-sorting), you can use the following:
```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    inv_sorted = inverse.argsort(stable=True)
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    return unique, inverse, counts, index
```
* For PyTorch Versions 1.12 and below (i.e., incapable of stable-sorting), you can use the following:
```python
def unique(x, dim=0):
    unique, inverse, counts = torch.unique(x, dim=dim, 
        sorted=True, return_inverse=True, return_counts=True)
    decimals = torch.arange(inverse.numel(), device=inverse.device) / inverse.numel()
    inv_sorted = (inverse+decimals).argsort()
    tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
    index = inv_sorted[tot_counts]
    return unique, inverse, counts, index
```

Hope this helps.

**Comment by wjaekim** [comment]
As ehsansaleh mentioned above, this issue is also discussed in #36748.

For an even larger band-aid (^^), I coded a function that covers a wider range of use cases. Here's a quote from #36748.

> I needed the indices not only of unique slices but also of _last occurring_ unique slices. Even further, some use cases required such indices for unsorted, unique elements. In fact, unsorted outputs are not currently supported in PyTorch whenever dim is not None (even if unique indices are not requested). If you consider inverse indices, this problem gets more complicated.

> So, I extended the function to cover all possible use cases. Of course, this should be regarded as a temporary patch until these are hopefully implemented as PyTorch's backend code. But I doubt it---some of these functionalities are still not adopted in Numpy.

>I made sure output tensors' shapes are identical to torch.unique's in all cases. Also, this extended function applies only if necessary; otherwise, torch.unique is used.

Here is the function (w/ PyTorch Versions 1.13 and above assumed):

```python
def uniqueXT(x, sorted=True, return_index=False, return_inverse=False, return_counts=False,
             occur_last=False, dim=None):
    if return_index or (not sorted and dim is not None):
        unique, inverse, counts = torch.unique(x, sorted=True,
            return_inverse=True, return_counts=True, dim=dim)
        inv_sorted, inv_argsort = inverse.flatten().sort(stable=True)

        if occur_last and return_index:
            tot_counts = (inverse.numel() - 1 - 
                torch.cat((counts.new_zeros(1),
                counts.flip(dims=[0]).cumsum(dim=0)))[:-1].flip(dims=[0]))
        else:
            tot_counts = torch.cat((counts.new_zeros(1), counts.cumsum(dim=0)))[:-1]
        
        index = inv_argsort[tot_counts]
        
        if not sorted:
            index, idx_argsort = index.sort()
            unique = (unique[idx_argsort] if dim is None else
                torch.index_select(unique, dim, idx_argsort))
            if return_inverse:
                idx_tmp = idx_argsort.argsort()
                inverse.flatten().index_put_((inv_argsort,), idx_tmp[inv_sorted])
            if return_counts:
                counts = counts[idx_argsort]

        ret = (unique,)
        if return_index:
            ret += (index,)
        if return_inverse:
            ret += (inverse,)
        if return_counts:
            ret += (counts,)
        return ret if len(ret)>1 else ret[0]
    
    else:
        return torch.unique(x, sorted=sorted, return_inverse=return_inverse,
            return_counts=return_counts, dim=dim)
```
>
